File: Trabalho_docker_5/graphql_app.py
import strawberry
from fastapi import FastAPI
from strawberry.fastapi import GraphQLRouter
from typing import List
from db import SessionLocal, Usuario as DbUsuario, Musica as DbMusica, Playlist as DbPlaylist

@strawberry.type
class Musica:
    id: int
    nome: str
    artista: str

@strawberry.type
class Playlist:
    id: int
    nome: str

@strawberry.type
class Usuario:
    id: int
    nome: str
    idade: int

@strawberry.type
class Query:
    @strawberry.field
    def usuarios(self) -> List[Usuario]:
        db = SessionLocal()
        users = db.query(DbUsuario).all()
        db.close()
        return [Usuario(id=u.id, nome=u.nome, idade=u.idade) for u in users]

    @strawberry.field
    def musicas(self) -> List[Musica]:
        db = SessionLocal()
        musicas = db.query(DbMusica).all()
        db.close()
        return [Musica(id=m.id, nome=m.nome, artista=m.artista) for m in musicas]

schema = strawberry.Schema(query=Query)
graphql_app = GraphQLRouter(schema)

app = FastAPI(title="GraphQL Streaming API")
app.include_router(graphql_app, prefix="/graphql")

File: Trabalho_docker_5/typescript/src/graphql_app.ts
import { ApolloServer, gql } from 'apollo-server';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

const typeDefs = gql`
  type Usuario { id: Int!, nome: String!, idade: Int }
  type Musica { id: Int!, nome: String!, artista: String! }
  type Playlist { id: Int!, nome: String!, usuario_id: Int! }

  type Query {
    usuarios: [Usuario]
    musicas: [Musica]
    playlistsUsuario(userId: Int!): [Playlist]
    musicasPlaylist(playlistId: Int!): [Musica]
    playlistsPorMusica(musicaId: Int!): [Playlist]
  }
`;

const resolvers = {
  Query: {
    usuarios: () => prisma.usuarios.findMany(),
    musicas: () => prisma.musicas.findMany(),
    playlistsUsuario: (_: any, args: { userId: number }) => prisma.playlists.findMany({ where: { usuario_id: args.userId } }),
    musicasPlaylist: async (_: any, args: { playlistId: number }) => {
        const data = await prisma.playlist_musica.findMany({ where: { playlist_id: args.playlistId }, include: { musicas: true } });
        return data.map((d: any) => d.musicas);
    },
    playlistsPorMusica: async (_: any, args: { musicaId: number }) => {
        const data = await prisma.playlist_musica.findMany({ where: { musica_id: args.musicaId }, include: { playlists: true } });
        return data.map((d: any) => d.playlists);
    }
  }
};

const server = new ApolloServer({ typeDefs, resolvers });

server.listen({ port: 9001 }).then(({ url }) => {
  console.log(`TS GraphQL pronta em ${url}`);
});